src/graphene.cpp

```diff
@@ -0,0 +1,1471 @@
+    {
+        dosMan.Misbehaving(pfrom, 100);
+        return error("Incorrectly constructed get_grblocktx received.  Banning peer=%s", pfrom->GetLogName());
+    }
+
+    // We use MSG_TX here even though we refer to blockhash because we need to track
+    // how many grblocktx requests we make in case of DOS
+    CInv inv(MSG_TX, grapheneRequestBlockTx.blockhash);
+    LOG(GRAPHENE, "Received get_grblocktx for %s peer=%s\n", inv.hash.ToString(), pfrom->GetLogName());
+
+    // Check for Misbehaving and DOS
+    // If they make more than 20 requests in 10 minutes then disconnect them
+    {
+        LOCK(cs_vNodes);
+        if (pfrom->nGetGrapheneBlockTxLastTime <= 0)
+            pfrom->nGetGrapheneBlockTxLastTime = GetTime();
+        uint64_t nNow = GetTime();
+        pfrom->nGetGrapheneBlockTxCount *=
+            std::pow(1.0 - 1.0 / 600.0, (double)(nNow - pfrom->nGetGrapheneBlockTxLastTime));
+        pfrom->nGetGrapheneBlockTxLastTime = nNow;
+        pfrom->nGetGrapheneBlockTxCount += 1;
+        LOG(GRAPHENE, "nGetGrapheneTxCount is %f\n", pfrom->nGetGrapheneBlockTxCount);
+        if (pfrom->nGetGrapheneBlockTxCount >= 20)
+        {
+            dosMan.Misbehaving(pfrom, 100); // If they exceed the limit then disconnect them
+            return error("DOS: Misbehaving - requesting too many grblocktx: %s\n", inv.hash.ToString());
+        }
+    }
+
+    {
+        LOCK(cs_main);
+        std::vector<CTransaction> vTx;
+        BlockMap::iterator mi = mapBlockIndex.find(inv.hash);
+        if (mi == mapBlockIndex.end())
+        {
+            dosMan.Misbehaving(pfrom, 20);
+            return error("Requested block is not available");
+        }
+        else
+        {
+            CBlock block;
+            const Consensus::Params &consensusParams = Params().GetConsensus();
+            if (!ReadBlockFromDisk(block, (*mi).second, consensusParams))
+            {
+                // We do not assign misbehavior for not being able to read a block from disk because we already
+                // know that the block is in the block index from the step above. Secondly, a failure to read may
+                // be our own issue or the remote peer's issue in requesting too early.  We can't know at this point.
+                return error("Cannot load block from disk -- Block txn request possibly received before assembled");
+            }
+            else
+            {
+                for (auto &tx : block.vtx)
+                {
+                    uint64_t cheapHash = tx->GetHash().GetCheapHash();
+
+                    if (grapheneRequestBlockTx.setCheapHashesToRequest.count(cheapHash))
+                        vTx.push_back(*tx);
+                }
+            }
+        }
+        CGrapheneBlockTx grapheneBlockTx(grapheneRequestBlockTx.blockhash, vTx);
+        pfrom->PushMessage(NetMsgType::GRAPHENETX, grapheneBlockTx);
+        pfrom->blocksSent += 1;
```

Comment:
seems like the wrong statistic.  We are sending missing transactions, not a block.

Reply:
Fixed with 4934022.

Also, I initially copied this code from thinblocks.cpp, which uses blocksSent as well. Maybe we should make a similar change there?